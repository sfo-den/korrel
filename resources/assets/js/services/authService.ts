import { merge } from 'lodash'
import { http, localStorageService } from '@/services'
import { userStore } from '@/stores'

export interface UpdateCurrentProfileData {
  current_password: string | null
  name: string
  email: string
  avatar?: string
  new_password?: string
}

interface CompositeToken {
  'audio-token': string
  'token': string
}

const API_TOKEN_STORAGE_KEY = 'api-token'
const AUDIO_TOKEN_STORAGE_KEY = 'audio-token'

export const authService = {
  async login (email: string, password: string) {
    const token = await http.post<CompositeToken>('me', { email, password })

    this.setAudioToken(token['audio-token'])
    this.setApiToken(token.token)
  },

  async logout () {
    await http.delete('me')
    this.destroy()
  },

  getProfile: async () => await http.get<User>('me'),

  updateProfile: async (data: UpdateCurrentProfileData) => {
    merge(userStore.current, (await http.put<User>('me', data)))
  },

  getApiToken: () => localStorageService.get(API_TOKEN_STORAGE_KEY),

  hasApiToken () {
    return Boolean(this.getApiToken())
  },

  setApiToken: (token: string) => localStorageService.set(API_TOKEN_STORAGE_KEY, token),

  destroy: () => {
    localStorageService.remove(API_TOKEN_STORAGE_KEY)
    localStorageService.remove(AUDIO_TOKEN_STORAGE_KEY)
  },

  setAudioToken: (token: string) => localStorageService.set(AUDIO_TOKEN_STORAGE_KEY, token),

  getAudioToken: () => {
    // for backward compatibility, we first try to get the audio token, and fall back to the (full-privileged) API token
    return localStorageService.get(AUDIO_TOKEN_STORAGE_KEY) || localStorageService.get(API_TOKEN_STORAGE_KEY)
  }
}
