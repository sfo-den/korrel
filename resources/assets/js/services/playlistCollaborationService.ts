import { http } from '@/services'

export const playlistCollaborationService = {
  async createInviteLink (playlist: Playlist) {
    if (playlist.is_smart) {
      throw Error('Smart playlists are not collaborative.')
    }

    const token = (await http.post<{ token: string }>(`playlists/${playlist.id}/collaborators/invite`)).token
    return `${window.location.origin}/#/playlist/collaborate/${token}`
  },

  async acceptInvite (token: string) {
    return http.post<Playlist>(`playlists/collaborators/accept`, { token })
  }
}
