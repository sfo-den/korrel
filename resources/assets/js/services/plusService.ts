import { http } from '@/services'

export const plusService = {
  activateLicense: async (key: string) => {
    return await http.post('licenses/activate', { key })
  }
}
