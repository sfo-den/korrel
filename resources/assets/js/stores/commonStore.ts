import isMobile from 'ismobilejs'
import { reactive } from 'vue'
import { http } from '@/services'
import { playlistFolderStore, playlistStore, preferenceStore, queueStore, settingStore, themeStore, userStore } from '.'

interface CommonStoreState {
  allows_download: boolean
  cdn_url: string
  current_user: User
  current_version: string
  latest_version: string
  koel_plus: {
    active: boolean
    short_key: string | null
    customer_name: string | null
    customer_email: string | null
    product_id: string
  }
  media_path_set: boolean
  playlists: Playlist[]
  playlist_folders: PlaylistFolder[]
  queue_state: QueueState,
  settings: Settings
  song_count: number,
  song_length: number,
  uses_i_tunes: boolean
  uses_last_fm: boolean
  uses_spotify: boolean
  uses_you_tube: boolean
  users: User[]
}

export const commonStore = {
  state: reactive<CommonStoreState>({
    allows_download: false,
    cdn_url: '',
    current_user: undefined as unknown as User,
    current_version: '',
    koel_plus: {
      active: false,
      short_key: null,
      customer_name: null,
      customer_email: null,
      product_id: ''
    },
    latest_version: '',
    media_path_set: false,
    playlists: [],
    playlist_folders: [],
    settings: {} as Settings,
    uses_i_tunes: false,
    uses_last_fm: false,
    uses_spotify: false,
    users: [],
    uses_you_tube: false,
    song_count: 0,
    song_length: 0,
    queue_state: {
      type: 'queue-states',
      songs: [],
      current_song: null,
      playback_position: 0
    }
  }),

  async init () {
    Object.assign(this.state, await http.get<CommonStoreState>('data'))

    // Always disable YouTube integration on mobile.
    this.state.uses_you_tube = this.state.uses_you_tube && !isMobile.phone

    // If this is a new user, initialize his preferences to be an empty object.
    this.state.current_user.preferences = this.state.current_user.preferences || {}

    userStore.init(this.state.current_user)
    preferenceStore.init(this.state.current_user.preferences)
    playlistStore.init(this.state.playlists)
    playlistFolderStore.init(this.state.playlist_folders)
    settingStore.init(this.state.settings)
    queueStore.init(this.state.queue_state)
    themeStore.init()

    return this.state
  }
}
