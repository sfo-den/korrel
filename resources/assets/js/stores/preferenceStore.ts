import { reactive, ref } from 'vue'
import { http, localStorageService } from '@/services'

interface Preferences extends Record<string, any> {
  volume: number
  show_now_playing_notification: boolean
  repeat_mode: RepeatMode
  confirm_before_closing: boolean
  equalizer: EqualizerPreset,
  artists_view_mode: ArtistAlbumViewMode | null,
  albums_view_mode: ArtistAlbumViewMode | null,
  transcode_on_mobile: boolean
  support_bar_no_bugging: boolean
  show_album_art_overlay: boolean
  lyrics_zoom_level: number | null
  theme?: Theme['id'] | null
  visualizer?: Visualizer['id'] | null
  active_extra_panel_tab: ExtraPanelTab | null
  make_uploads_public: boolean
}

const preferenceStore = {
  initialized: ref(false),

  state: reactive<Preferences>({
    volume: 7,
    show_now_playing_notification: true,
    repeat_mode: 'NO_REPEAT',
    confirm_before_closing: false,
    equalizer: {
      name: 'Default',
      preamp: 0,
      gains: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    artists_view_mode: null,
    albums_view_mode: null,
    transcode_on_mobile: false,
    support_bar_no_bugging: false,
    show_album_art_overlay: true,
    lyrics_zoom_level: 1,
    theme: null,
    visualizer: 'default',
    active_extra_panel_tab: null,
    make_uploads_public: false
  }),

  init (preferences: Preferences): void {
    Object.assign(this.state, preferences)
    this.setupProxy()

    this.initialized.value = true
  },

  /**
   * Proxy the state properties, so that each can be directly accessed using the key.
   */
  setupProxy () {
    Object.keys(this.state).forEach(key => {
      Object.defineProperty(this, key, {
        get: (): any => this.get(key),
        set: (value: any): void => this.set(key, value),
        configurable: true
      })
    })
  },

  set (key: keyof Preferences, value: any) {
    if (this.state[key] === value) return

    this.state[key] = value
    http.silently.patch('me/preferences', { key, value })
  },

  get (key: string) {
    return this.state?.[key]
  }
}

const exported = preferenceStore as unknown as Omit<typeof preferenceStore, 'setupProxy'> & Preferences

export { exported as preferenceStore }
