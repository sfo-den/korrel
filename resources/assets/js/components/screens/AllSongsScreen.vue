<template>
  <section id="songsWrapper">
    <ScreenHeader :layout="headerLayout">
      All Songs
      <ControlsToggle v-model="showingControls" />

      <template #thumbnail>
        <ThumbnailStack :thumbnails="thumbnails" />
      </template>

      <template v-if="totalSongCount" #meta>
        <span>{{ pluralize(totalSongCount, 'song') }}</span>
        <span>{{ totalDuration }}</span>
      </template>

      <template #controls>
        <div class="controls">
          <SongListControls
            v-if="totalSongCount && (!isPhone || showingControls)"
            @play-all="playAll"
            @play-selected="playSelected"
          />
          <label class="own-songs-toggle text-secondary" v-if="isPlus">
            <CheckBox v-model="ownSongsOnly" />
            <span>Own songs only</span>
          </label>
        </div>
      </template>
    </ScreenHeader>

    <SongListSkeleton v-if="showSkeletons" />
    <template v-else>
      <SongList
        v-if="songs.length > 0"
        ref="songList"
        @sort="sort"
        @scroll-breakpoint="onScrollBreakpoint"
        @press:enter="onPressEnter"
        @scrolled-to-end="fetchSongs"
      />
      <ScreenEmptyState v-else>
        <template #icon>
          <Icon :icon="faVolumeOff" />
        </template>
        Your library is empty.
        <a
          role="button"
          class="d-block secondary"
          v-if="isPlus && ownSongsOnly"
          @click.prevent="showSongsFromOthers"
        >
          Show public songs from other users?
        </a>
      </ScreenEmptyState>
    </template>
  </section>
</template>

<script lang="ts" setup>
import { faVolumeOff } from '@fortawesome/free-solid-svg-icons'
import { computed, ref, toRef, watch } from 'vue'
import { logger, pluralize, secondsToHumanReadable } from '@/utils'
import { commonStore, queueStore, songStore } from '@/stores'
import { localStorageService, playbackService } from '@/services'
import { useMessageToaster, useKoelPlus, useRouter, useSongList } from '@/composables'

import ScreenHeader from '@/components/ui/ScreenHeader.vue'
import SongListSkeleton from '@/components/ui/skeletons/SongListSkeleton.vue'
import CheckBox from '@/components/ui/CheckBox.vue'
import ScreenEmptyState from '@/components/ui/ScreenEmptyState.vue'

const totalSongCount = toRef(commonStore.state, 'song_count')
const totalDuration = computed(() => secondsToHumanReadable(commonStore.state.song_length))

const {
  SongList,
  SongListControls,
  ControlsToggle,
  ThumbnailStack,
  headerLayout,
  thumbnails,
  songs,
  songList,
  duration,
  showingControls,
  isPhone,
  onPressEnter,
  playSelected,
  onScrollBreakpoint
} = useSongList(toRef(songStore.state, 'songs'))

const { toastError } = useMessageToaster()
const { go, onScreenActivated } = useRouter()
const { isPlus } = useKoelPlus()

let initialized = false
const loading = ref(false)
let sortField: SongListSortField = 'title' // @todo get from query string
let sortOrder: SortOrder = 'asc'

const page = ref<number | null>(1)
const moreSongsAvailable = computed(() => page.value !== null)
const showSkeletons = computed(() => loading.value && songs.value.length === 0)

const ownSongsOnly = ref(isPlus.value ? Boolean(localStorageService.get('own-songs-only')) : false)

watch(ownSongsOnly, async value => {
  localStorageService.set('own-songs-only', value)
  page.value = 1
  songStore.state.songs = []

  await fetchSongs()
})

const sort = async (field: SongListSortField, order: SortOrder) => {
  page.value = 1
  songStore.state.songs = []
  sortField = field
  sortOrder = order

  await fetchSongs()
}

const showSongsFromOthers = async () => {
  ownSongsOnly.value = false
  await fetchSongs()
}

const fetchSongs = async () => {
  if (!moreSongsAvailable.value || loading.value) return

  loading.value = true

  try {
    page.value = await songStore.paginate({
      sort: sortField,
      order: sortOrder,
      page: page.value!,
      own_songs_only: ownSongsOnly.value
    })
  } catch (error) {
    toastError('Failed to load songs.')
    logger.error(error)
  } finally {
    loading.value = false
  }
}

const playAll = async (shuffle: boolean) => {
  if (shuffle) {
    await queueStore.fetchRandom()
  } else {
    await queueStore.fetchInOrder(sortField, sortOrder)
  }

  go('queue')
  await playbackService.playFirstInQueue()
}

onScreenActivated('Songs', async () => {
  if (!initialized) {
    initialized = true
    await fetchSongs()
  }
})
</script>

<style lang="scss" scoped>
.controls {
  width: 100%;
  min-height: 32px; // prevent shrinking causing the jumping effect
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: 1rem;

  .collapsed & {
    width: auto;
  }

  .own-songs-toggle {
    display: inline-flex;
    align-items: center;
    font-size: 1rem;
  }
}
</style>
