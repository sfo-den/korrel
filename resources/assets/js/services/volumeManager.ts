import { ref } from 'vue'
import { preferenceStore } from '@/stores'

export class VolumeManager {
  private input!: HTMLInputElement
  public volume = ref(0)

  public init (input: HTMLInputElement) {
    this.input = input
    this.set(preferenceStore.volume)
  }

  public get () {
    return this.volume.value
  }

  public set (volume: number) {
    this.volume.value = volume
    this.input.value = String(volume)
  }

  public mute () {
    this.set(0)
  }

  public unmute () {
    this.set(preferenceStore.volume)
  }
}

export const volumeManager = new VolumeManager()
