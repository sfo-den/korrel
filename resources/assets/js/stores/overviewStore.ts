import { reactive, watch } from 'vue'
import { http } from '@/services'
import { songStore } from '@/stores/songStore'
import { albumStore } from '@/stores/albumStore'
import { artistStore } from '@/stores/artistStore'
import { recentlyPlayedStore } from '@/stores'

export const overviewStore = {
  state: reactive({
    recentlyPlayed: [] as Song[],
    recentlyAddedSongs: [] as Song[],
    recentlyAddedAlbums: [] as Album[],
    mostPlayedSongs: [] as Song[],
    mostPlayedAlbums: [] as Album[],
    mostPlayedArtists: [] as Artist[]
  }),

  async fetch () {
    const resource = await http.get<{
      most_played_songs: Song[],
      most_played_albums: Album[],
      most_played_artists: Artist[],
      recently_added_songs: Song[],
      recently_added_albums: Album[],
      recently_played_songs: Song[],
    }>('overview')

    songStore.syncWithVault(resource.most_played_songs)
    albumStore.syncWithVault(resource.recently_added_albums)
    artistStore.syncWithVault(resource.most_played_artists)

    this.state.mostPlayedAlbums = albumStore.syncWithVault(resource.most_played_albums)
    this.state.mostPlayedArtists = artistStore.syncWithVault(resource.most_played_artists)
    this.state.recentlyAddedSongs = songStore.syncWithVault(resource.recently_added_songs)
    this.state.recentlyAddedAlbums = albumStore.syncWithVault(resource.recently_added_albums)

    recentlyPlayedStore.excerptState.playables = songStore.syncWithVault(resource.recently_played_songs)

    this.refreshPlayStats()
  },

  refreshPlayStats () {
    this.state.mostPlayedSongs = songStore.getMostPlayed(7)
    this.state.recentlyPlayed = recentlyPlayedStore.excerptState.playables.filter(
      ({ deleted, play_count }) => !deleted && play_count > 0
    )
  }
}
