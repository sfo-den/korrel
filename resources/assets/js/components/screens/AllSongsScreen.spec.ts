import { expect, it } from 'vitest'
import factory from '@/__tests__/factory'
import UnitTestCase from '@/__tests__/UnitTestCase'
import { commonStore, queueStore, songStore } from '@/stores'
import { screen, waitFor } from '@testing-library/vue'
import { playbackService } from '@/services'
import AllSongsScreen from './AllSongsScreen.vue'

new class extends UnitTestCase {
  private async renderComponent () {
    commonStore.state.song_count = 420
    commonStore.state.song_length = 123_456
    songStore.state.songs = factory<Song>('song', 20)
    const fetchMock = this.mock(songStore, 'paginate').mockResolvedValue(2)

    this.router.$currentRoute.value = {
      screen: 'Songs',
      path: '/songs'
    }

    const rendered = this.render(AllSongsScreen, {
      global: {
        stubs: {
          SongList: this.stub('song-list')
        }
      }
    })

    await waitFor(() => expect(fetchMock).toHaveBeenCalledWith({
      sort: 'title',
      order: 'asc',
      page: 1,
      own_songs_only: false
    }))

    return [rendered, fetchMock] as const
  }

  protected test () {
    it('renders', async () => {
      const [{ html }] = await this.renderComponent()
      await waitFor(() => expect(html()).toMatchSnapshot())
    })

    it('shuffles', async () => {
      const queueMock = this.mock(queueStore, 'fetchRandom')
      const playMock = this.mock(playbackService, 'playFirstInQueue')
      const goMock = this.mock(this.router, 'go')
      await this.renderComponent()

      await this.user.click(screen.getByTitle('Shuffle all. Press Alt/âŒ¥ to change mode.'))

      await waitFor(() => {
        expect(queueMock).toHaveBeenCalled()
        expect(playMock).toHaveBeenCalled()
        expect(goMock).toHaveBeenCalledWith('queue')
      })
    })

    it('renders in Plus edition', async () => {
      this.enablePlusEdition()

      const [{ html }, fetchMock] = await this.renderComponent()
      await waitFor(() => expect(html()).toMatchSnapshot())

      await this.user.click(screen.getByLabelText('Own songs only'))
      await waitFor(() => expect(fetchMock).toHaveBeenCalledWith({
        sort: 'title',
        order: 'asc',
        page: 1,
        own_songs_only: true
      }))
    })
  }
}
