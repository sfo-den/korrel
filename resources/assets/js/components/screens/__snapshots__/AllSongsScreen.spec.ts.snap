// Vitest Snapshot v1

exports[`renders 1`] = `
<section data-v-8ea4eaa5="" id="songsWrapper">
  <header data-v-5691beb5="" data-v-8ea4eaa5="" class="screen-header expanded">
    <aside data-v-5691beb5="" class="thumbnail-wrapper">
      <div data-v-55bfc268="" data-v-8ea4eaa5="" data-v-5691beb5-s="" class="thumbnail-stack single" style="background-image: url(undefined/resources/assets/img/covers/default.svg);"><span data-v-55bfc268="" data-testid="thumbnail"></span></div>
    </aside>
    <main data-v-5691beb5="">
      <div data-v-5691beb5="" class="heading-wrapper">
        <h1 data-v-5691beb5="" class="name"> All Songs
          <!--v-if-->
        </h1><span data-v-5691beb5="" class="meta text-secondary"><span data-v-8ea4eaa5="" data-v-5691beb5-s="">420 songs</span><span data-v-8ea4eaa5="" data-v-5691beb5-s="">34 hr 17 min</span></span>
      </div>
      <div data-v-8ea4eaa5="" data-v-5691beb5-s="" class="controls">
        <div data-v-d396e0d2="" data-v-8ea4eaa5="" data-v-5691beb5-s="" class="song-list-controls" data-testid="song-list-controls">
          <div data-v-d396e0d2="" class="wrapper"><span data-v-e884c19a="" data-v-d396e0d2="" class="btn-group" uppercased=""><button data-v-e368fe26="" data-v-d396e0d2="" class="btn-shuffle-all" data-testid="btn-shuffle-all" orange="" title="Shuffle all. Press Alt/⌥ to change mode."><br data-v-d396e0d2="" data-testid="Icon" icon="[object Object]" fixed-width=""> All </button><!--v-if--><!--v-if--><!--v-if--></span>
            <!--v-if-->
            <!--v-if-->
          </div>
          <div data-v-d396e0d2="" class="menu-wrapper">
            <div data-v-42061e3e="" data-v-d396e0d2="" class="add-to" data-testid="add-to-menu" tabindex="0">
              <section data-v-42061e3e="" class="existing-playlists">
                <p data-v-42061e3e="">Add 0 songs to</p>
                <ul data-v-42061e3e="">
                  <li data-v-42061e3e="" data-testid="queue" tabindex="0">Queue</li>
                  <li data-v-42061e3e="" class="favorites" data-testid="add-to-favorites" tabindex="0"> Favorites </li>
                </ul>
              </section><button data-v-e368fe26="" data-v-42061e3e="" transparent="">New Playlist…</button>
            </div>
          </div>
        </div>
        <!--v-if-->
      </div>
    </main>
  </header><br data-v-8ea4eaa5="" data-testid="song-list">
</section>
`;

exports[`renders in Plus edition 1`] = `
<section data-v-8ea4eaa5="" id="songsWrapper">
  <header data-v-5691beb5="" data-v-8ea4eaa5="" class="screen-header expanded">
    <aside data-v-5691beb5="" class="thumbnail-wrapper">
      <div data-v-55bfc268="" data-v-8ea4eaa5="" data-v-5691beb5-s="" class="thumbnail-stack single" style="background-image: url(undefined/resources/assets/img/covers/default.svg);"><span data-v-55bfc268="" data-testid="thumbnail"></span></div>
    </aside>
    <main data-v-5691beb5="">
      <div data-v-5691beb5="" class="heading-wrapper">
        <h1 data-v-5691beb5="" class="name"> All Songs
          <!--v-if-->
        </h1><span data-v-5691beb5="" class="meta text-secondary"><span data-v-8ea4eaa5="" data-v-5691beb5-s="">420 songs</span><span data-v-8ea4eaa5="" data-v-5691beb5-s="">34 hr 17 min</span></span>
      </div>
      <div data-v-8ea4eaa5="" data-v-5691beb5-s="" class="controls">
        <div data-v-d396e0d2="" data-v-8ea4eaa5="" data-v-5691beb5-s="" class="song-list-controls" data-testid="song-list-controls">
          <div data-v-d396e0d2="" class="wrapper"><span data-v-e884c19a="" data-v-d396e0d2="" class="btn-group" uppercased=""><button data-v-e368fe26="" data-v-d396e0d2="" class="btn-shuffle-all" data-testid="btn-shuffle-all" orange="" title="Shuffle all. Press Alt/⌥ to change mode."><br data-v-d396e0d2="" data-testid="Icon" icon="[object Object]" fixed-width=""> All </button><!--v-if--><!--v-if--><!--v-if--></span>
            <!--v-if-->
            <!--v-if-->
          </div>
          <div data-v-d396e0d2="" class="menu-wrapper">
            <div data-v-42061e3e="" data-v-d396e0d2="" class="add-to" data-testid="add-to-menu" tabindex="0">
              <section data-v-42061e3e="" class="existing-playlists">
                <p data-v-42061e3e="">Add 0 songs to</p>
                <ul data-v-42061e3e="">
                  <li data-v-42061e3e="" data-testid="queue" tabindex="0">Queue</li>
                  <li data-v-42061e3e="" class="favorites" data-testid="add-to-favorites" tabindex="0"> Favorites </li>
                </ul>
              </section><button data-v-e368fe26="" data-v-42061e3e="" transparent="">New Playlist…</button>
            </div>
          </div>
        </div><label data-v-8ea4eaa5="" data-v-5691beb5-s="" class="own-songs-toggle text-secondary"><span data-v-b5259680="" data-v-8ea4eaa5="" data-v-5691beb5-s="" class=""><input data-v-b5259680="" type="checkbox"></span><span data-v-8ea4eaa5="" data-v-5691beb5-s="">Own songs only</span></label>
      </div>
    </main>
  </header><br data-v-8ea4eaa5="" data-testid="song-list">
</section>
`;
